pkg/util/globalerror/grpc.go

```diff
@@ -82,47 +81,31 @@ func (e *grpcContextError) GRPCStatus() *grpcstatus.Status {
 // and grpc/status packages, but which preserve the original error
 // semantics.
 type ErrorWithStatus struct {
-	UnderlyingErr error
-	Status        *status.Status
+	Causes []error
+	Status *status.Status
 }
 
-func NewErrorWithGRPCStatus(originalErr error, code codes.Code, details *mimirpb.ErrorDetails) ErrorWithStatus {
-	stat := status.New(code, originalErr.Error())
-	if details != nil {
-		if statWithDetails, err := stat.WithDetails(details); err == nil {
-			return ErrorWithStatus{
-				UnderlyingErr: originalErr,
-				Status:        statWithDetails,
-			}
-		}
-	}
-	return ErrorWithStatus{
-		UnderlyingErr: originalErr,
-		Status:        stat,
-	}
-}
-
-func (e ErrorWithStatus) Error() string {
+func (e *ErrorWithStatus) Error() string {
```

Comment:
pointer receivers on a method sometimes indicate that the type can be mutated by the method. Can't we get away with value receivers - `(e ErrorWithStatus)`

Reply:
Initially I implemented it as you suggested. The problem is that with a value receiver `ErrorWithStatus.Causes` was not correctly passed form one method to another. That's why I decided to implement it this way.

Reply:
When the values are pointers, then I think we hit this condition and the actual pointer addresses are compared instead of the errors https://cs.opensource.google/go/go/+/refs/tags/go1.22.3:src/errors/wrap.go;l=55-57

This became a problem as `Causes` became a slice and slices make structs not comparable. Which makes me have second thoughts about having a slice in the struct. Maybe we can use `fmt.Errorf("%w: %w", err, context.Canceled)` in this place and avoid needing to store a slice ourselves. Or simply store two booleans in `ErrorWithStatus` indicating whether the error is a context cancelation or a deadline exceeded or neither.

Reply:
Hi @dimitarvdimitrov , 
I have removed pointer receivers, and returned `ErrorWithStatus` from every function that used to return `*ErrorWithStatus`. I have also removed the `Causes []error` slice, and put back the previous `UnderlyingErr error`, but added another field `ctxErr err` that gets assigned a value only after `WrapGRPCErrorWithContextError ()` is called.

Similarly, `Unwrap() []error` returns `[]error{e.UnderlyingErr}` or `[]error{UnderlyingErr, ctxErr}` depending on whether `ctxErr` is set or not.